src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageUpdaterPreTFMStep.cs

```diff
@@ -0,0 +1,204 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Steps.Packages
+{
+    /// <summary>
+    /// Upgrade step that updates NuGet package references
+    /// to better work after upgrade. Packages references are
+    /// updated if the reference appears to be transitive (with
+    /// SDK style projects, only top-level dependencies are necessary
+    /// in the project file), if the package version doesn't
+    /// target a compatible .NET framework but a newer version does,
+    /// or if the package is explicitly mapped to an updated
+    /// NuGet package in a mapping configuration file.
+    /// </summary>
+    public class PackageUpdaterPreTFMStep : UpgradeStep
+    {
+        private const int MaxAnalysisIterations = 3;
+
+        private readonly IPackageRestorer _packageRestorer;
+        private readonly IEnumerable<IPackageReferencesAnalyzer> _packageAnalyzers;
+
+        private PackageAnalysisState? _analysisState;
+
+        public override string Description => "Update package references and remove transitive dependencies";
+
+        public override string Title => "Update Nuget Packages Before Updating the TFM";
```

Comment:
"Clean up NuGet package references" might be a better description to distinguish it from the later step because, theoretically, nothing should really be updated - we should just be removing unnecessary references and things like that at this point.

Reply:
Changed